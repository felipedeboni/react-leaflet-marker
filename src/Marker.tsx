import classNames from "classnames";
import React, {FC} from 'react';

import useLatitudeOffset from "./hooks/useLatitudeOffset";
import useMarkerRef from "./hooks/useMarkerRef";
import usePlacementAsPoint from "./hooks/usePlacementAsPoint";
import useUpdatePosition from "./hooks/useUpdatePosition";
import useZIndex from "./hooks/useZIndex";
import useZoomAnimation from "./hooks/useZoomAnimation";
import styles from './Marker.module.css';
import { EPlacement, IMarkerProps } from "./Marker.types";


const Marker: FC<IMarkerProps> = ({
    children,
    position,
    size,
    innerRef,
    riseOffset = 250,
    zIndexOffset = 0,
    interactive= false,
    riseOnHover = false,
    placement = EPlacement.top,
}) => {
    const markerRef = useMarkerRef(innerRef);
    const placementAsPoint = usePlacementAsPoint(placement, size);
    const latitudeOffset = useLatitudeOffset(placementAsPoint, position);
    const hasAnimation = useZoomAnimation(placementAsPoint, position, markerRef);
    const [ zIndex, domEvents ] = useZIndex(
        latitudeOffset,
        zIndexOffset,
        riseOffset,
        riseOnHover
    );
    useUpdatePosition(
        placementAsPoint,
        position,
        markerRef,
    );

    return (
        <div
            ref={markerRef}
            onMouseEnter={domEvents?.onMouseEnter}
            onMouseLeave={domEvents?.onMouseLeave}
            className={classNames(styles.wrapper, {
                [styles.wrapper_interactive]: interactive,
                ['leaflet-zoom-animated']: hasAnimation,
                ['leaflet-zoom-hide']: !hasAnimation,
            })}
            style={{
                zIndex,
                ...(size ? {
                    width: size[0],
                    height: size[0],
                } : null),
            }}
        >
            {children}
        </div>
    );
}

export default Marker;