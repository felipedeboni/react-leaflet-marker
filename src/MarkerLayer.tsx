import {useLeafletContext} from "@react-leaflet/core";
import React, {FC, ReactElement, useLayoutEffect, useRef} from 'react';

import {IMarkerProps} from "./Marker.types";
import {IMarkerLayerProps} from "./MarkerLayer.types";

const MarkerLayer: FC<IMarkerLayerProps> = ({
  children,
  pane = 'overlayPane'
}) => {
  const map = useLeafletContext().map;
  const layerRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    if (layerRef.current) {
      map.getPanes()[pane].appendChild(layerRef.current);
    }
  }, [map, pane]);

  return (
    <div
        ref={layerRef}
        className="leaflet-objects-pane leaflet-marker-pane"
    >
      {React.Children.map(
          children,
          (child) => React.cloneElement(
              child as ReactElement<IMarkerProps>,
          ))
      }
    </div>
  );
};

export default MarkerLayer;
