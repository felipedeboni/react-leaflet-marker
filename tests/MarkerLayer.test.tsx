import { useLeafletContext } from '@react-leaflet/core';
import { cleanup, render } from "@testing-library/react";
import React from 'react'

import Marker from '../src/Marker'
import MarkerLayer from '../src/MarkerLayer'

jest.mock('../src/Marker', () => function mockMarker(props: JSX.IntrinsicAttributes & React.ClassAttributes<HTMLDivElement> & React.HTMLAttributes<HTMLDivElement>) {return <div data-component="Marker" {...props} />});
jest.mock('@react-leaflet/core', () => {
    const pane = {
        appendChild: jest.fn(),
        removeChild: jest.fn(),
    };
    const map = {
        getPanes: () => ({
            overlayPane: pane,
            markerPane: pane,
        })
    };

    return {
        useLeafletContext: jest.fn(() => ({
            map
        }))
    };
});

describe('MarkerLayer', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    test('MarkerLayer add overlay', () => {
        const map = useLeafletContext().map;

        render(
            <MarkerLayer>
                <Marker position={{lat: 55.84294011297764, lng: 48.52798461914063}}>children</Marker>
            </MarkerLayer>
        );

        expect(map.getPanes().overlayPane.appendChild).toHaveBeenCalledTimes(1);
        expect(map.getPanes().overlayPane.removeChild).toHaveBeenCalledTimes(0);
    });

    test('MarkerLayer unmount', () => {
        const map = useLeafletContext().map;

        const { unmount } = render(
            <MarkerLayer pane="markerPane">
                <Marker position={{lat: 55.84294011297764, lng: 48.52798461914063}}>children</Marker>
            </MarkerLayer>
        );

        unmount();

        expect(map.getPanes().markerPane.appendChild).toHaveBeenCalledTimes(1);
        expect(map.getPanes().markerPane.removeChild).toHaveBeenCalledTimes(1);
    });


    test('MarkerLayer rerender', () => {
        const map = useLeafletContext().map;

        const { rerender } = render(
            <MarkerLayer>
                <Marker position={{lat: 55.84294011297764, lng: 48.52798461914063}}>children</Marker>
            </MarkerLayer>
        );
        expect(map.getPanes().overlayPane.appendChild).toHaveBeenCalledTimes(1);

        rerender(
            <MarkerLayer pane="markerPane">
                <Marker position={{lat: 55.84294011297764, lng: 48.52798461914063}}>children</Marker>
            </MarkerLayer>
        );
        expect(map.getPanes().overlayPane.removeChild).toHaveBeenCalledTimes(1);
        expect(map.getPanes().markerPane.appendChild).toHaveBeenCalledTimes(2);
    });

    test('Snapshots', () => {
        const { asFragment } = render(
            <MarkerLayer>
                <Marker position={{lat: 55.84294011297764, lng: 48.52798461914063}}>children 1</Marker>
                <Marker position={{lat: 52.84294011297764, lng: 48.52798461914063}}>children 2</Marker>
                <Marker position={{lat: 45.84294011297764, lng: 48.52798461914063}}>children 3</Marker>
            </MarkerLayer>
        )

        expect(asFragment()).toMatchSnapshot();
    });
});