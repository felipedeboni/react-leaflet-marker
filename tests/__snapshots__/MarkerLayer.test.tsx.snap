// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerLayer Snapshots 1`] = `
<DocumentFragment>
  <div
    class="leaflet-objects-pane leaflet-marker-pane"
  >
    <div
      data-component="Marker"
      position="[object Object]"
    >
      children 1
    </div>
    <div
      data-component="Marker"
      position="[object Object]"
    >
      children 2
    </div>
    <div
      data-component="Marker"
      position="[object Object]"
    >
      children 3
    </div>
  </div>
</DocumentFragment>
`;
